File: ios/Sources/RekognitionLivenessPlugin.swift
//
//  RekognitionLivenessPlugin.swift
//  tauri-plugin-rekognition-liveness
//
//  Hosts the AWS Amplify FaceLivenessDetectorView (SwiftUI) as a sibling of
//  the Tauri WKWebView (mirrors the official barcode-scanner plugin pattern).
//  Camera capture and the WebSocket streaming connection to AWS Rekognition
//  Streaming live entirely native-side — only `{ status, error? }` crosses
//  back through the Tauri IPC bridge.
//

import AVFoundation
import AWSPluginsCore
import FaceLiveness
import Foundation
import SwiftUI
import Tauri
import UIKit
import WebKit

class LivenessCredentialsArgs: Decodable {
    let accessKeyId: String
    let secretAccessKey: String
    let sessionToken: String
    let expiresAt: String
}

class LivenessDisplayTextArgs: Decodable {
    /// Overrides the SDK's "Center your face" prompt.
    let centerFace: String?
}

class DetectLivenessArgs: Decodable {
    let sessionId: String
    let region: String
    let credentials: LivenessCredentialsArgs
    /// "front" (default) or "back". Only takes effect when the session was
    /// created with `Settings.ChallengePreferences = [FaceMovementChallenge]`;
    /// the default `FaceMovementAndLightChallenge` forces front regardless.
    let camera: String?
    /// Optional UI-text overrides applied for the duration of this session.
    let displayText: LivenessDisplayTextArgs?
}

/// Maps platform-neutral display-text keys (the names exposed on the
/// `displayText` JS arg) to the localizable keys Amplify Liveness Swift looks
/// up via `NSLocalizedString(... bundle: .main, ...)`. Anything not in this
/// map is silently ignored (forward-compat with future plugin keys the
/// vendored SDK doesn't yet honour).
private let displayTextKeyMap: [String: String] = [
    "centerFace": "amplify_ui_liveness_center_your_face_text"
]

/// Process-wide override store consulted by the `Bundle.localizedString`
/// swizzle below. Empty when no liveness session is active → swizzle becomes
/// a no-op (just calls through to the original implementation).
private final class LivenessLocalizedOverrides {
    static let shared = LivenessLocalizedOverrides()
    private let lock = NSLock()
    private var values: [String: String] = [:]

    func set(_ overrides: [String: String]) {
        lock.lock(); defer { lock.unlock() }
        values = overrides
    }

    func clear() {
        lock.lock(); defer { lock.unlock() }
        values.removeAll()
    }

    func lookup(_ key: String) -> String? {
        lock.lock(); defer { lock.unlock() }
        return values[key]
    }
}

/// One-shot installer for the `Bundle.localizedString(forKey:value:table:)`
/// method swizzle. Without this, the Amplify SDK's `NSLocalizedString` calls
/// against `Bundle.main` only resolve from the host app's `Localizable.strings`
/// — runtime overrides have no path in. We swap the instance method for our
/// stub once at process start; the stub checks
/// `LivenessLocalizedOverrides.shared` first, then falls through to the
/// (now-renamed) original implementation. Idempotent via `dispatch_once`-style
/// static let.
private enum BundleLocalizedSwizzle {
    static let install: Void = {
        let cls: AnyClass = Bundle.self
        let original = #selector(Bundle.localizedString(forKey:value:table:))
        let replacement = #selector(Bundle.tplrl_localizedString(forKey:value:table:))
        guard
            let originalMethod = class_getInstanceMethod(cls, original),
            let replacementMethod = class_getInstanceMethod(cls, replacement)
        else { return }
        method_exchangeImplementations(originalMethod, replacementMethod)
    }()
}

extension Bundle {
    @objc fileprivate func tplrl_localizedString(
        forKey key: String,
        value: String?,
        table tableName: String?
    ) -> String {
        if let override = LivenessLocalizedOverrides.shared.lookup(key) {
            return override
        }
        // Post-swizzle this calls the *original* implementation thanks to
        // `method_exchangeImplementations` — naming notwithstanding.
        return self.tplrl_localizedString(forKey: key, value: value, table: tableName)
    }
}

class RekognitionLivenessPlugin: Plugin {
    private var savedInvoke: Invoke?
    private var hostingController: UIHostingController<AnyView>?
    private var willResignObserver: NSObjectProtocol?

    @objc public override func checkPermissions(_ invoke: Invoke) {
        invoke.resolve(["camera": currentCameraPermissionState()])
    }

    @objc public override func requestPermissions(_ invoke: Invoke) {
        let state = currentCameraPermissionState()
        if state == "prompt" {
            AVCaptureDevice.requestAccess(for: .video) { granted in
                invoke.resolve(["camera": granted ? "granted" : "denied"])
            }
        } else {
            invoke.resolve(["camera": state])
        }
    }

    @objc public func detectLiveness(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(DetectLivenessArgs.self)

        if savedInvoke != nil {
            invoke.reject("Another liveness session is already in progress")
            return
        }

        // iOS aborts the process with a runtime exception when
        // AVCaptureDevice.requestAccess is called without an
        // NSCameraUsageDescription string in the host's Info.plist. Fail fast
        // with a clearer error before triggering the OS prompt.
        let entry = Bundle.main.infoDictionary?["NSCameraUsageDescription"] as? String
        if entry == nil || entry?.isEmpty == true {
            invoke.resolve([
                "status": "error",
                "error": [
                    "code": "MissingPermissionsString",
                    "message": "NSCameraUsageDescription is not set in the host app's Info.plist",
                ],
            ])
            return
        }

        savedInvoke = invoke

        // Mirror the Android pattern: gate the Amplify UI mount on a runtime
        // camera permission check. Amplify's FaceLivenessDetector grabs the
        // camera the instant it composes; without permission it short-circuits
        // to `onError(.cameraPermissionDenied)` and the user sees a confusing
        // "go to Settings" alert. Resolve permission first, then mount.
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            DispatchQueue.main.async { [weak self] in self?.mountLivenessUi(args: args) }
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
                DispatchQueue.main.async {
                    guard let self = self else { return }
                    if granted {
                        self.mountLivenessUi(args: args)
                    } else {
                        self.resolve(
                            status: "failed",
                            error: ("CameraPermissionDenied",
                                    "Camera permission is required for liveness detection"))
                    }
                }
            }
        case .denied, .restricted:
            fallthrough
        @unknown default:
            resolve(
                status: "failed",
                error: ("CameraPermissionDenied",
                        "Camera permission is required for liveness detection"))
        }
    }

    private func currentCameraPermissionState() -> String {
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            return "granted"
        case .denied, .restricted:
            return "denied"
        case .notDetermined:
            return "prompt"
        @unknown default:
            return "prompt"
        }
    }

    private func mountLivenessUi(args: DetectLivenessArgs) {
        guard let viewController = self.manager.viewController,
              let webView = viewController.view.subviews.first(where: { $0 is WKWebView }) as? WKWebView,
              let parent = webView.superview else {
            resolve(status: "error", error: ("WebViewUnavailable",
                                            "Unable to locate the host WKWebView to mount the liveness UI."))
            return
        }

        // Install Bundle swizzle once, then load any caller-supplied UI-text
        // overrides for this session. Cleared in `teardown()` so the swizzle
        // becomes a no-op while no liveness UI is on screen.
        _ = BundleLocalizedSwizzle.install
        if let displayText = args.displayText {
            var overrides: [String: String] = [:]
            if let v = displayText.centerFace, let key = displayTextKeyMap["centerFace"] {
                overrides[key] = v
            }
            LivenessLocalizedOverrides.shared.set(overrides)
        } else {
            LivenessLocalizedOverrides.shared.clear()
        }

        let credentialsProvider = StaticAWSCredentialsProvider(args: args.credentials)

        // The session's challenge type (FaceMovementAndLight vs FaceMovement)
        // is fixed when the backend calls `CreateFaceLivenessSession`. We only
        // get to control which camera the SDK should use *if* the session
        // ends up running FaceMovement — FaceMovementAndLight forces the
        // front camera (the screen has to face the user for the light
        // challenge to mean anything).
        let livenessCamera: LivenessCamera = (args.camera?.lowercased() == "back") ? .back : .front
        let challengeOptions = ChallengeOptions(
            faceMovementChallengeOption: FaceMovementChallengeOption(camera: livenessCamera)
        )

        let livenessView = FaceLivenessDetectorView(
            sessionID: args.sessionId,
            credentialsProvider: credentialsProvider,
            region: args.region,
            challengeOptions: challengeOptions,
            isPresented: Binding<Bool>(
                get: { true },
                set: { _ in }
            ),
            onCompletion: { [weak self] result in
                switch result {
                case .success:
                    self?.resolve(status: "success", error: nil)
                case .failure(let err):
                    self?.resolve(status: "failed", error: ("FaceLivenessError", "\(err)"))
                }
            }
        )

        let host = UIHostingController(rootView: AnyView(livenessView))
        host.view.frame = parent.bounds
        host.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        host.view.backgroundColor = UIColor.clear
        parent.insertSubview(host.view, aboveSubview: webView)
        hostingController = host

        // Backgrounding mid-session → cancelled.
        willResignObserver = NotificationCenter.default.addObserver(
            forName: UIApplication.willResignActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.resolve(status: "cancelled", error: nil)
        }
    }

    private func resolve(status: String, error: (String, String)?) {
        guard let invoke = savedInvoke else { return }
        savedInvoke = nil

        var payload: [String: Any] = ["status": status]
        if let (code, message) = error {
            payload["error"] = ["code": code, "message": message]
        }
        invoke.resolve(payload)

        DispatchQueue.main.async { [weak self] in
            self?.teardown()
        }
    }

    private func teardown() {
        hostingController?.view.removeFromSuperview()
        hostingController = nil

        if let observer = willResignObserver {
            NotificationCenter.default.removeObserver(observer)
            willResignObserver = nil
        }

        // Drop any per-session UI-text overrides — the next NSLocalizedString
        // call from anywhere in the host app falls through to the SDK / app
        // bundle as if the swizzle weren't there.
        LivenessLocalizedOverrides.shared.clear()
    }
}

/// Concrete `AWSTemporaryCredentials` carrier for STS-issued creds passed in
/// from JS. Amplify's `AWSCredentialsProvider` expects values matching the
/// `AWSCredentials` (or `AWSTemporaryCredentials`) protocol, not a struct
/// type — so we vend a minimal struct that conforms.
private struct StaticTemporaryCredentials: AWSTemporaryCredentials {
    let accessKeyId: String
    let secretAccessKey: String
    let sessionToken: String
    let expiration: Date
}

/// Static credentials provider wrapping the JS-supplied STS credentials.
/// Plugin doesn't refresh — caller must mint creds with TTL ≥ ~5 min so they
/// outlive the Liveness session (max ~3 min).
private struct StaticAWSCredentialsProvider: AWSCredentialsProvider {
    let args: LivenessCredentialsArgs

    func fetchAWSCredentials() async throws -> AWSCredentials {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let expiration = formatter.date(from: args.expiresAt)
            ?? ISO8601DateFormatter().date(from: args.expiresAt)
            ?? Date().addingTimeInterval(15 * 60)
        return StaticTemporaryCredentials(
            accessKeyId: args.accessKeyId,
            secretAccessKey: args.secretAccessKey,
            sessionToken: args.sessionToken,
            expiration: expiration
        )
    }
}

@_cdecl("init_plugin_rekognition_liveness")
func initPlugin() -> Plugin {
    return RekognitionLivenessPlugin()
}

File: android/src/main/java/RekognitionLivenessPlugin.kt
package app.tauri.rekognitionliveness

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import android.content.pm.PackageManager
import android.content.res.Resources
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.view.ViewGroup
import android.webkit.WebView
import android.widget.FrameLayout
import androidx.activity.ComponentActivity
import androidx.activity.OnBackPressedCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.AbstractComposeView
import androidx.compose.ui.platform.LocalContext
import androidx.core.content.ContextCompat
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.Permission
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import aws.smithy.kotlin.runtime.time.Instant
import com.amplifyframework.auth.AWSCredentials
import com.amplifyframework.auth.AWSCredentialsProvider
import com.amplifyframework.auth.AWSTemporaryCredentials
import com.amplifyframework.auth.AuthException
import com.amplifyframework.core.Consumer
import com.amplifyframework.ui.liveness.ui.Camera as AmplifyCamera
import com.amplifyframework.ui.liveness.ui.ChallengeOptions
import com.amplifyframework.ui.liveness.ui.FaceLivenessDetector
import com.amplifyframework.ui.liveness.ui.LivenessChallenge

@InvokeArg
class LivenessCredentialsArgs {
    lateinit var accessKeyId: String
    lateinit var secretAccessKey: String
    lateinit var sessionToken: String
    lateinit var expiresAt: String
}

@InvokeArg
class LivenessDisplayTextArgs {
    /** Overrides the SDK's "Center your face" prompt. */
    var centerFace: String? = null
}

@InvokeArg
class DetectLivenessArgs {
    lateinit var sessionId: String
    lateinit var region: String
    lateinit var credentials: LivenessCredentialsArgs

    /**
     * "front" (default) or "back". Only takes effect when the session was
     * created with `Settings.ChallengePreferences = [FaceMovementChallenge]`;
     * the default `FaceMovementAndLightChallenge` forces front regardless.
     * `var?` rather than `lateinit` because the JS caller may omit it.
     */
    var camera: String? = null

    /**
     * Optional UI-text overrides applied for the duration of this session.
     * Lets the consuming app re-word prompts that don't fit its context (e.g.
     * a rear-camera-at-the-gate flow re-wording "Center your face").
     */
    var displayText: LivenessDisplayTextArgs? = null
}

/**
 * Maps platform-neutral display-text keys (the names exposed on the
 * `displayText` JS arg) to Amplify Liveness Android string resource names.
 * Resource ids are looked up at runtime via `Resources.getIdentifier` —
 * library AAR `R.string` ids are remapped to consumer-app ids during
 * resource-merging at build time, so the consumer sees the same id the SDK
 * sees, and overriding `getString(id)` for that id wins everywhere.
 */
private val DISPLAY_TEXT_KEY_MAP = mapOf(
    "centerFace" to "amplify_ui_liveness_get_ready_center_face_label",
)

/**
 * Drop-in [Resources] subclass that returns caller-supplied strings for a
 * fixed set of resource ids and falls through to the wrapped resources for
 * everything else. Constructed once per session.
 *
 * The `Resources(AssetManager, DisplayMetrics, Configuration)` constructor
 * has been deprecated since API 18 with the guidance "apps should not
 * construct Resources directly", but Android has never offered a public
 * alternative for *subclassing* Resources — every framework / Compose
 * helper that needs to intercept string lookup ends up here. Suppression is
 * the canonical fix; behaviour is stable across all supported API levels.
 */
@Suppress("DEPRECATION")
private class StringOverrideResources(
    private val base: Resources,
    private val overrides: Map<Int, String>,
) : Resources(base.assets, base.displayMetrics, base.configuration) {
    override fun getString(id: Int): String =
        overrides[id] ?: super.getString(id)

    override fun getString(id: Int, vararg formatArgs: Any?): String =
        overrides[id]?.let { String.format(it, *formatArgs) }
            ?: super.getString(id, *formatArgs)

    override fun getText(id: Int): CharSequence =
        overrides[id] ?: super.getText(id)

    override fun getText(id: Int, def: CharSequence?): CharSequence =
        overrides[id] ?: super.getText(id, def)
}

/**
 * [ContextWrapper] that returns the [StringOverrideResources] above. We pass
 * an instance of this through Compose's `LocalContext` so the bundled
 * `FaceLivenessDetector` resolves `stringResource(...)` against our overrides
 * without the SDK needing to know the override mechanism exists.
 */
private class StringOverrideContext(
    base: Context,
    private val overrideResources: StringOverrideResources,
) : ContextWrapper(base) {
    override fun getResources(): Resources = overrideResources
}

/**
 * Build the override-id map from the caller's [LivenessDisplayTextArgs]. Any
 * key whose Android resource id can't be resolved (e.g. plugin built against
 * a newer SDK that renamed the string) is silently dropped.
 */
private fun resolveDisplayTextOverrides(
    context: Context,
    args: LivenessDisplayTextArgs?,
): Map<Int, String> {
    if (args == null) return emptyMap()
    val pairs = listOfNotNull(
        args.centerFace?.let { DISPLAY_TEXT_KEY_MAP["centerFace"] to it },
    )
    if (pairs.isEmpty()) return emptyMap()
    val pkg = context.packageName
    return pairs.mapNotNull { (name, value) ->
        if (name == null) return@mapNotNull null
        val id = context.resources.getIdentifier(name, "string", pkg)
        if (id == 0) null else id to value
    }.toMap()
}

/**
 * Hosts the Amplify [FaceLivenessDetector] composable as a sibling of the
 * Tauri WebView (mirrors the official barcode-scanner plugin pattern). All
 * camera capture and the WebSocket streaming connection to AWS Rekognition
 * Streaming live entirely native-side — only `{ status, error? }` crosses
 * back through the Tauri IPC bridge.
 */
@TauriPlugin(
    permissions = [
        Permission(strings = [Manifest.permission.CAMERA], alias = "camera"),
    ],
)
class RekognitionLivenessPlugin(private val activity: Activity) : Plugin(activity) {
    private lateinit var webView: WebView
    private var savedInvoke: Invoke? = null
    private var pendingArgs: DetectLivenessArgs? = null
    private var hostView: AbstractComposeView? = null
    private var webViewBackground: Drawable? = null
    private var backCallback: OnBackPressedCallback? = null
    private var lifecycleObserver: DefaultLifecycleObserver? = null
    private var cameraPermissionLauncher: ActivityResultLauncher<Array<String>>? = null

    override fun load(webView: WebView) {
        super.load(webView)
        this.webView = webView

        // Tauri 2.11.0 made `app.tauri.plugin.PluginManager` a singleton object
        // but its `requestPermissionsLauncher` (registered in `onActivityCreate`)
        // is never invoked by the runtime — so `Plugin.requestPermissionForAlias`
        // throws `lateinit property requestPermissionsLauncher has not been
        // initialized` on first use. We register our own ActivityResultLauncher
        // against the host activity's registry to bypass the broken wiring.
        // Tauri 2.10 doesn't have this issue (PluginManager was an instance
        // there, and TauriActivity constructed it during class init).
        (activity as? ComponentActivity)?.let { componentActivity ->
            cameraPermissionLauncher = componentActivity.activityResultRegistry.register(
                "rekognition-liveness:cameraPermission",
                ActivityResultContracts.RequestMultiplePermissions(),
            ) { result ->
                onCameraPermissionResult(result[Manifest.permission.CAMERA] == true)
            }
        }
    }

    @Command
    fun detectLiveness(invoke: Invoke) {
        if (savedInvoke != null) {
            invoke.reject("Another liveness session is already in progress")
            return
        }

        val args = invoke.parseArgs(DetectLivenessArgs::class.java)
        savedInvoke = invoke
        pendingArgs = args

        // Amplify's FaceLivenessDetector touches the camera the moment it
        // composes; if runtime CAMERA permission isn't granted we skip
        // straight to onError("CameraPermissionDeniedException"). Gate the
        // mount on a runtime permission check, requesting via the OS dialog
        // when needed (mirrors official tauri-plugin-barcode-scanner).
        if (isCameraPermissionGranted()) {
            activity.runOnUiThread { mountLivenessUi(args) }
        } else {
            val launcher = cameraPermissionLauncher
            if (launcher != null) {
                launcher.launch(arrayOf(Manifest.permission.CAMERA))
            } else {
                resolve(
                    "failed",
                    LivenessFailure(
                        code = "CameraPermissionDenied",
                        message = "Host activity is not a ComponentActivity; cannot request camera permission",
                    ),
                )
            }
        }
    }

    private fun onCameraPermissionResult(granted: Boolean) {
        // A `requestPermissions` JS call has priority over `detectLiveness`'s
        // implicit prompt — it expects a `{camera: ...}` payload back, not the
        // detection-result shape.
        pendingPermissionInvoke?.let { invoke ->
            pendingPermissionInvoke = null
            invoke.resolve(JSObject().apply {
                put("camera", if (granted) "granted" else "denied")
            })
            return
        }

        val args = pendingArgs
        if (granted && args != null) {
            activity.runOnUiThread { mountLivenessUi(args) }
        } else {
            resolve(
                "failed",
                LivenessFailure(
                    code = "CameraPermissionDenied",
                    message = "Camera permission is required for liveness detection",
                ),
            )
        }
    }

    private fun isCameraPermissionGranted(): Boolean =
        ContextCompat.checkSelfPermission(activity, Manifest.permission.CAMERA) ==
            PackageManager.PERMISSION_GRANTED

    private fun cameraPermissionStateString(): String = when {
        isCameraPermissionGranted() -> "granted"
        // No way to reliably distinguish "denied forever" from "prompt" without
        // the activity context to call shouldShowRequestPermissionRationale,
        // and the result's only useful for caller UX hints. Keep it simple.
        else -> "prompt"
    }

    @Command
    override fun checkPermissions(invoke: Invoke) {
        invoke.resolve(JSObject().apply { put("camera", cameraPermissionStateString()) })
    }

    @Command
    override fun requestPermissions(invoke: Invoke) {
        if (isCameraPermissionGranted()) {
            invoke.resolve(JSObject().apply { put("camera", "granted") })
            return
        }
        val launcher = cameraPermissionLauncher
        if (launcher == null) {
            invoke.resolve(JSObject().apply { put("camera", "denied") })
            return
        }
        // Hijack the existing camera launcher: when the user responds, resolve
        // the JS-facing requestPermissions invoke instead of running the
        // detect-liveness mount path. We swap the callback by tagging
        // `pendingArgs == null` (no detection in flight) and intercepting the
        // result via a one-shot side-channel.
        pendingPermissionInvoke = invoke
        launcher.launch(arrayOf(Manifest.permission.CAMERA))
    }

    private var pendingPermissionInvoke: Invoke? = null

    private fun mountLivenessUi(args: DetectLivenessArgs) {
        val parent = webView.parent as ViewGroup
        webViewBackground = webView.background
        webView.setBackgroundColor(Color.TRANSPARENT)

        // We avoid `ComposeView(activity).apply { setContent { ... } }` because
        // the lambda's `Function0` bytecode does not match the runtime
        // `ComposeView.setContent(Function2<Composer, Int, Unit>)` ABI inside
        // the Tauri-bundled library module — Tauri's auto-generated consumer
        // app declares no `composeOptions`, and the cross-module compose
        // compiler handling silently drops our lambda transform. An
        // `AbstractComposeView` subclass with overridden `Content()` resolves
        // the call via virtual dispatch, sidestepping the lambda ABI entirely.
        val composeView = LivenessHostView(activity, args, this::resolve).apply {
            layoutParams = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
            )
        }
        parent.addView(composeView)
        hostView = composeView

        // Hardware-back must end the session, not pop the host activity.
        // onBackPressedDispatcher + lifecycle live on ComponentActivity, not
        // plain Activity — guard accordingly.
        (activity as? ComponentActivity)?.let { componentActivity ->
            backCallback = object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    resolve("cancelled", null)
                }
            }.also {
                componentActivity.onBackPressedDispatcher.addCallback(componentActivity, it)
            }

            lifecycleObserver = object : DefaultLifecycleObserver {
                override fun onPause(owner: LifecycleOwner) {
                    if (savedInvoke != null) resolve("cancelled", null)
                }
            }.also {
                componentActivity.lifecycle.addObserver(it)
            }
        }
    }

    @Synchronized
    private fun resolve(status: String, failure: LivenessFailure?) {
        val invoke = savedInvoke ?: return
        savedInvoke = null
        pendingArgs = null

        val payload = JSObject().apply {
            put("status", status)
            failure?.let {
                put(
                    "error",
                    JSObject().apply {
                        put("code", it.code)
                        put("message", it.message)
                    },
                )
            }
        }
        invoke.resolve(payload)
        teardown()
    }

    private fun teardown() {
        activity.runOnUiThread {
            hostView?.let { (it.parent as? ViewGroup)?.removeView(it) }
            hostView = null
            webViewBackground?.let { webView.background = it }
            webView.setBackgroundColor(Color.WHITE)
            webViewBackground = null

            backCallback?.remove()
            backCallback = null

            (activity as? ComponentActivity)?.let { componentActivity ->
                lifecycleObserver?.let { componentActivity.lifecycle.removeObserver(it) }
            }
            lifecycleObserver = null
        }
    }

    private data class LivenessFailure(val code: String, val message: String)

    /**
     * Hosts the Compose `FaceLivenessDetector` via an `AbstractComposeView`
     * subclass instead of `ComposeView.setContent { ... }`. The override goes
     * through virtual dispatch, so the runtime call site is fixed at the
     * override and unaffected by cross-module Compose compiler quirks that
     * leave the `setContent` lambda untransformed (`Function0` vs the expected
     * composer-aware `Function2`).
     */
    private class LivenessHostView(
        context: Context,
        private val args: DetectLivenessArgs,
        private val onResult: (status: String, failure: LivenessFailure?) -> Unit,
    ) : AbstractComposeView(context) {
        @Composable
        override fun Content() {
            MaterialTheme {
                Surface(modifier = Modifier.fillMaxSize()) {
                    // The session's challenge type (FaceMovementAndLight vs
                    // FaceMovement) is fixed when the backend calls
                    // `CreateFaceLivenessSession`; we don't pick it here. We
                    // do, however, supply the camera the SDK should use *if*
                    // the session ends up running FaceMovement — the
                    // FaceMovementAndLight path ignores any camera hint and
                    // forces front (the screen has to face the user for the
                    // light challenge to be meaningful).
                    val camera: AmplifyCamera = when (args.camera?.lowercase()) {
                        "back" -> AmplifyCamera.Back
                        else -> AmplifyCamera.Front
                    }
                    val challengeOptions = ChallengeOptions(
                        faceMovementAndLight = LivenessChallenge.FaceMovementAndLight,
                        faceMovement = LivenessChallenge.FaceMovement(camera = camera),
                    )

                    // Wrap LocalContext so the SDK's `stringResource(...)`
                    // calls resolve against caller-supplied overrides for the
                    // string ids we care about. No-op when displayText is
                    // unset / empty — Compose just sees the original context.
                    val baseContext = LocalContext.current
                    val overrideContext = remember(args.displayText) {
                        val overrides =
                            resolveDisplayTextOverrides(baseContext, args.displayText)
                        if (overrides.isEmpty()) baseContext
                        else StringOverrideContext(
                            baseContext,
                            StringOverrideResources(baseContext.resources, overrides),
                        )
                    }

                    CompositionLocalProvider(LocalContext provides overrideContext) {
                        FaceLivenessDetector(
                            sessionId = args.sessionId,
                            region = args.region,
                            credentialsProvider = StaticCredentialsProvider(args.credentials),
                            challengeOptions = challengeOptions,
                            onComplete = { onResult("success", null) },
                            onError = { error ->
                                onResult(
                                    "failed",
                                    LivenessFailure(
                                        code = error::class.simpleName ?: "FaceLivenessError",
                                        message = error.message ?: error.toString(),
                                    ),
                                )
                            },
                        )
                    }
                }
            }
        }
    }

    /**
     * Wraps the JS-supplied STS credentials in the callback-based provider
     * shape that Amplify Liveness expects (`com.amplifyframework.auth.AWSCredentialsProvider`,
     * not the Smithy coroutine variant). Plugin doesn't refresh — caller
     * must mint creds whose TTL outlives the session (≥5 minutes).
     */
    private class StaticCredentialsProvider(
        private val args: LivenessCredentialsArgs,
    ) : AWSCredentialsProvider<AWSCredentials> {
        override fun fetchAWSCredentials(
            onSuccess: Consumer<AWSCredentials>,
            onError: Consumer<AuthException>,
        ) {
            val expiration: Instant = try {
                Instant.fromIso8601(args.expiresAt)
            } catch (_: Exception) {
                // Fallback: 15 minutes from now if the ISO string didn't parse —
                // close to STS GetSessionToken's default TTL so behaviour is benign.
                Instant.fromEpochSeconds(System.currentTimeMillis() / 1000L + 15L * 60L)
            }
            // AWSTemporaryCredentials is a concrete final class, not an
            // interface — instantiate via its constructor.
            val creds = AWSTemporaryCredentials(
                accessKeyId = args.accessKeyId,
                secretAccessKey = args.secretAccessKey,
                sessionToken = args.sessionToken,
                expiration = expiration,
            )
            onSuccess.accept(creds)
        }
    }
}
